### ISSUE TITLE
Folded multiply optimizes poorly on WebAssembly

### ISSUE THREAD

**Issue by jsheard** [tkaitchuck/aHash#76]
https://rust.godbolt.org/z/dPYf4M

The fallback hash is designed around the assumption that the 128-bit multiply will be optimized into a 64-bit multiply with 128-bit output, but when compiling for WebAssembly it turns into a call to `__multi3` which emulates a full 128-bit multiply.

I think this is an inherent limitation of WebAssembly rather than a codegen issue, as far as I can tell there's no way to get the upper part of a 64-bit multiply in wasm. A secondary fallback that avoids 128-bit math might be beneficial for targets like this.

**Comment by jsheard** [comment]
A quick test to see how much of a performance cliff this causes

```rust
use ahash::AHasher;
use fnv::FnvHasher;
use rand::prelude::*;
use std::hash::{Hash, Hasher};
use std::time::Instant;

fn bench(name: &str, data: &[u64], mut hasher: impl Hasher) {
    let start = Instant::now();
    data.hash(&mut hasher);
    let hash = hasher.finish();
    let time = start.elapsed();

    println!("{}: {:?} ({})", name, time, hash);
}

fn main() {
    let mut data = vec![0u64; 8_000_000];
    thread_rng().fill(data.as_mut_slice());

    bench("ahash", &data, AHasher::default());
    bench("  fnv", &data, FnvHasher::default());
}
```

Native (no AES):
```
$ cargo run --release
ahash: 7.0502ms (16905656033929795835)
  fnv: 60.6068ms (8858245011109644555)
```

WebAssembly:
```
$ cargo build --target=wasm32-wasi --release
$ wasmer target\wasm32-wasi\release\wasmtest.wasm
ahash: 94ms (1894176816958052022)
  fnv: 109ms (18003293742428832362)
```

That's still acceptable performance on wasm, but AHashes usual performance edge is almost completely negated

**Comment by syrusakbary** [comment]
Have you tried running with the LLVM compiler, runtime speed performance might be increased!

```
$ wasmer run target\wasm32-wasi\release\wasmtest.wasm --llvm
```

**Comment by jsheard** [comment]
```
$ wasmer run target\wasm32-wasi\release\wasmtest.wasm --llvm
error: failed to run `target\wasm32-wasi\release\wasmtest.wasm`
╰─> 1: The `llvm` compiler is not included in this binary.
```

I'm using the Windows binary from https://github.com/wasmerio/wasmer/releases/tag/1.0.2

Does the LLVM backend only come with the Linux/OSX versions?

```
$ wasmtime target\wasm32-wasi\release\wasmtest.wasm
ahash: 37.4199ms (3833488243860778540)
  fnv: 82.4112ms (14874192434453256082)
```

wasmtime shows better performance overall, but ahash is still only 2x faster than FNV compared to 10x faster in native code

**Comment by syrusakbary** [comment]
Which version of Wasmer are you using, and in which system?
The prebuilt binaries should all include llvm except for Windows

**Comment by jsheard** [comment]
It occurred to me that hashing the entire dataset at once isn't really representative of how ahash/fnv would behave in a hashmap keyed by u64, where each hasher would only be used to hash a single value. Reworked the benchmark:

```rust
use ahash::RandomState;
use fnv::FnvBuildHasher;
use rand::prelude::*;
use std::hash::{BuildHasher, Hash, Hasher};
use std::time::Instant;

fn bench(name: &str, data: &[u64], builder: impl BuildHasher) {
    let mut total = 0u64;

    let start = Instant::now();
    for n in data.iter() {
        let mut hasher = builder.build_hasher();
        n.hash(&mut hasher);
        let hash = hasher.finish();

        // side effect to keep the hash from being optimized away
        total = total.wrapping_add(hash);
    }
    let time = start.elapsed();

    println!("{}: {:?} ({})", name, time, total);
}

fn main() {
    let mut data = vec![0u64; 8_000_000];
    thread_rng().fill(data.as_mut_slice());

    bench("ahash", &data, RandomState::default());
    bench("fnv1a", &data, FnvBuildHasher::default());
}
```

```
$ cargo run --release
ahash: 16.2001ms (2842195821360615028)
fnv1a: 21.3877ms (8444818633369249413)

$ cargo build --target=wasm32-wasi --release
$ wasmtime target\wasm32-wasi\release\wasmtest.wasm
ahash: 131.5265ms (12202533187417786512)
fnv1a: 34.5373ms (7541511931142809863)
```

It's actually worse than I thought, ahash goes from a slight lead over fnv in native code to being nearly 4x slower in wasm

**Comment by syrusakbary** [comment]
I was curious and I just tested times with the latest Wasmer in the new macOS M1.

Here's my timings:

### Wasmer (LLVM + Native)
```
➜  aHash git:(master) ✗ wasmer target/wasm32-wasi/release/example.wasm --llvm --native
ahash: 34.327ms (10030330502678093573)
fnv1a: 15.481ms (16252436043041238450)
````

### Wasmer (Cranelift + JIT - default)

```
➜  aHash git:(master) ✗ wasmer target/wasm32-wasi/release/example.wasm --cranelift
ahash: 42.439ms (15588686967232338973)
fnv1a: 25.012ms (10240259937822909892)
````

### Native binary

```
➜  aHash git:(master) ✗ cargo run --release
[...]
    Finished release [optimized] target(s) in 0.03s
     Running `target/release/example`
ahash: 5.835041ms (11922277158906542966)
fnv1a: 30.808083ms (6734640267817530018)
```

**Comment by jsheard** [comment]
FNV is faster in Wasmer than in a native binary on the M1? That's bizarre, I wonder what's going on there.

Nevertheless, ahash still struggles when compiled to WebAssembly.

**Comment by syrusakbary** [comment]
> FNV is faster in Wasmer than in a native binary on the M1? That's bizarre, I wonder what's going on there.

Yeah, I'm not 100% sure on why is that.
We have seen similar speedups in other benchmarks due to 32-bit pointers in Wasm instead of 64-bit ones in Native... that might be it?

> Nevertheless, ahash still struggles when compiled to WebAssembly.

Why you think is that? Is `ahash` using custom-encoded asm instructions when executed natively?

**Comment by jsheard** [comment]
> Why you think is that? Is `ahash` using custom-encoded asm instructions when executed natively?

AHash can optionally use AES crypto instructions to accelerate hashing on x86 targets, but all of my timings are taken with that feature disabled so the same portable fallback algorithm is used in both native and wasm tests. As I said in the OP, I think the problem is that the fallback algorithm assumes the processor can produce a 128-bit result from multiplying two 64-bit integers together, but WebAssembly doesn't support that operation and Rust has to generate very inefficient code to work around it.

**Comment by tkaitchuck** [comment]
I had hoped that LLVM would be able to work out that it only needed to perform a 64bit multiply. 
I have created the https://github.com/tkaitchuck/aHash/tree/miri branch to try out experimental options.

**Comment by jsheard** [comment]
LLVM can work it out, but there isn't necessarily a clean way to lower that to machine code

I checked all of the targets available on Godbolt to see which ones have fast `folded_multiply`

### Fast (1-2 mul instructions)

- `x86_64`
- `aarch64`
- `mips64`
- `powerpc64`
- `riscv64gc`
- `s390x`

### Not ideal (~10 math instructions)

- `i686`

### Slow (calls `__multi3`)

- `arm` (all 32-bit variants)
- `thumb` (all variants)
- `mips`
- `powerpc`
- `sparc64`
- `wasm32`

Many of those aren't particularly relevant anymore, but 32-bit ARM and Thumb are still around in embedded devices

**Comment by tkaitchuck** [comment]
The merged change creates a code path that should be good but is definitely not great, as it requires 4 multiplies per 64 bits of input. I am creating a new issue to track optimizing this: https://github.com/tkaitchuck/aHash/issues/79

**Comment by tkaitchuck** [comment]
The fix is included in version 0.7.2

**Comment by jsheard** [comment]
Thanks!
